### PR TITLE
Excel file support + 

### PR DESCRIPTION
**Pull request by dkrasner**
### Main Points ###

This PR adds excel (xlsx, xls) file support for both upload and download. It is based on the [sheet-js](https://docs.sheetjs.com/docs/getting-started/) library, which seems to be pretty active. In the process, I've added a [dialog](https://github.com/dkrasner/worksheet/compare/daniel-xlsx?expand=1#diff-d63b3e5baa408a9520b175842c16e48cf325e2556b82434ad808f0d17bf7a135R1203) element which is the interface to 

1. choosing which type of file to download (excel or csv). Arguably I could have done with 2 separate download icons but I couldn't find corresponding csv/excel ones in tabler and the dialog gives you the option to download both at the same time..
Screen Shot 2022-09-21 at 12 37 07 PM
2. choosing which sheet you want to upload, since we don't support multiple sheets/tabs (although i could see how to do this in worksheet down the line...). Again the dialog allows you to choose different sheets before committing. 

Screen Shot 2022-09-21 at 12 36 51 PM

Sheet names are preserved in both the worksheet and on download. 

A few things to note:
1. the sheet-js library does support csv's, although this is relatively new, but there were some minor issues and I decided to leave things as is. There is also a chance we'll have to drop it for another library because of...
2. dates dates.... uggh so excel natively supports dates as we know, but when you transfer from one software to another there is no guarantee that the same string representation will appear. The default for sheet-js is to display a numerical representation (although this doesn't seem to be seconds from epoch...), so when someone has "1/1/2022" they see some number on import - this is confusing. There are various [options](https://docs.sheetjs.com/docs/api/parse-options/) and after a while playing around I left it with [this](https://github.com/dkrasner/worksheet/compare/daniel-xlsx?expand=1#diff-d63b3e5baa408a9520b175842c16e48cf325e2556b82434ad808f0d17bf7a135R1147). Unfortunately this converts to a sheet-js default string representation. I think there might be a way to get around this by digging into the sheet-js worksheet data object and seeing if the original string rep is carried along, but I decided to kick the can down the road here.

### CODE REVIEW

File: js/Worksheet.js [hunk 1]
```diff
@@ -1087,6 +1135,120 @@ class Worksheet extends HTMLElement {
         return CSVParser.unparse(data);
     }
 
+    fromExcel(aString, rABS, fileName, ftype){
```
Comment (darth-cheney) [inline js/Worksheet.js:1138]: One issue I've encountered here is that -- unlike `fromCSV` -- this function does not update the size of the dataFrame of the current Worksheet. If the incoming data is bigger than the current dataFrame an error is thrown and nothing is loaded:

Screen Shot 2022-09-22 at 7 18 55 AM

**Comment by darth-cheney** [comment]
Aside from the one inline comment above, experimenting with this brought a couple of things to mind:
  
* Eventually our mappings will need to translate to some server-side capabilities (for batch processing). These will need to take into account the source file type, in this case Excel, which will need to include information about which sheet(s) of the workbook to process
* Unclear where this capability belongs. It is nice to have the `fromExcel` and `toExcel` right on the Worksheet element, of course, but it is becoming quite large. TBD.
* Right now we interpret all cell contents as strings. Is this going to be a problem? Do we need to rethink how data is "stored" in each cell?

**Review by darth-cheney — CHANGES_REQUESTED** [review]

File: js/Worksheet.js (hunk 1 above)
Comment (dkrasner) [reply js/Worksheet.js:1146]: @darth-cheney right, I forgot about this. There was a related issue with fromCSV. There is no guarantee in exports from MS/OpenOffice that the underlying nd-array (for .csv/.xlsx) is uniform, ie has the same length arrays/rows. 

From what I can see this is almost never the case, and what you see in sheet-js is an nd-array where each row-array is the minimum length necessary for the corresponding row, like the column-array length is the min for the colum, ie a maximally conservative approach. 

So if you were to upload a csv or excel with row data lengths [2, 10] then current dataFrame.corner.x = 2 and an error was thrown. 

There is also another [issue](https://github.com/darth-cheney/ap-sheet/issues/31). If you set the dataFrame size smaller than the size of the frame in view everything breaks. Similarly if you try to resize a sheet outside of the dataFrame size everything breaks. 

The [temp solution just pushed](https://github.com/dkrasner/worksheet/pull/39/commits/b27e2a3d9d35ffa7b4570552570951b0648ff008) is not going to solve the problem. But I think this should be solved at the sheet level, ie the sheet should be able to take whatever data is given it and grow when asked to etc. Worksheet doesn't need to know how sheet handles data, it's just a wrapper.

**Comment by dkrasner** [comment]
> Aside from the one inline comment above, experimenting with this brought a couple of things to mind:
> 
>     * Eventually our mappings will need to translate to some server-side capabilities (for batch processing). These will need to take into account the source file type, in this case Excel, which will need to include information about which sheet(s) of the workbook to process
> 
>     * Unclear where this capability belongs. It is nice to have the `fromExcel` and `toExcel` right on the Worksheet element, of course, but it is becoming quite large. TBD.
> 
we need to see what limits we are talking about here @ApproximateIdentity do you have any stats on this?

>     * Right now we interpret all cell contents as strings. Is this going to be a problem? Do we need to rethink how data is "stored" in each cell?
at the moment we'll leave everything as a string. If we need to deal with types (numerical, date etc) it's another bowl of soup. Hopefully those sorts of calculations will be left to python/R for them but I am probably wrong here.

**Comment by ApproximateIdentity** [comment]
Sorry no when I started messing with the JL server setup and ran into the issues I forgot I was planning on working through those stats. Obviously don't need JL's servers for that. I'll get it worked out on my end and get some numbers back to you

**Comment by ApproximateIdentity** [comment]
@dkrasner I can't seem to come up with any way to kill the setup by file sizes. Uploading 150mb csvs and 75mb excel files (had issues creating larger ones to due excel column limits) did work. I did have some issues trying to redownload them where it wasn't really working right, but it was hard to see what the issue was. I'm having trouble finding an error.

So basically I don't think that the actual webserver setup is an issue here. There are some things that we might want to consider:

1. How many temporary files are created and where? I.e. if there are excel files created in the jlapp container are they ever saved there on disk? Do they get saved to different names so there could be a space leak?
2. Assuming nothing is ever saved anywhere but the actual webdav location (I think this is true), are there issues on the server with memory running out? I.e. during some sort of temporary conversion are the files created in memory that could blow things out?

But maybe most importantly, I think we might want to initially set some file conditions that we want to guarantee we support. I.e. maybe we say the following:

1. We test to make sure we can support 50mb csvs and excel files.
2. We test to make sure we can support up to 1000 columns.
3. We test to make sure we can support up to 100000 rows.

These numbers are totally arbitrary, but if we set something it's a little easier to think things through. (Obviously 2. and 3. multiply so are limited by 1.)

Anyway I so far haven't hit any issues with the server. My browser got up to close to 10gb of memory, but the server itself seemed to have a very low memory footprint throughout.

**Pull request merged**